Only buffer one item at a time, prevents unbounded memory usage

xmakro opened the pull request:
Resolves https://github.com/jonhoo/async-bincode/issues/11

jonhoo reviewed (approved):
Excellent, thanks for going the extra mile!

jonhoo commented:
Ah, you may also need to update the tests in `src/lib.rs` so that they use a `BufWriter`, since currently they end up running _very_ slowly.

jonhoo commented:
Oh, just FYI, I also updated the CI configuration on `main` (now renamed from `master`), so you may want to rebase on top of that (but it's not necessary if you don't feel like it :) )

xmakro commented:
Thank you Jon. I updated branch and fixed a missing poll_write inside poll_flush. I added BufStream to the bigger tokio test, but not to the futures-io one as it would make the code more complicated since they only have separate BufWriter/BufReader and not the BufStream. All tests run in 0.22s in fastbuild and 0.1s in release mode, hope that is fine. Please take another look before merging and thank you very much for the quick turn around!

jonhoo commented on src/lib.rs:
```diff
@@ -137,14 +137,15 @@ mod tokio_tests {
-            let (r, w) = stream.tcp_split();
+            let stream =
+                AsyncBincodeStream::<_, usize, usize, _>::from(BufStream::new(stream)).for_async();
+            let (w, r) = stream.split();
```
Oof, the need to use the (much less efficient) `.split` here instead of `.tcp_split` is unfortunate. I think what this really implies is that we should have `AsyncBincodeStream` support keeping the reader and writer separately so that we can split _before_ constructing the `AsyncBincodeStream`. That's probably a change for down the line instead though.

jonhoo commented:
Released in 0.7.1 — thanks for taking the time to go through all the steps of this!
